<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>InPhaDel by l337x911</title>
    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/github-dark.css">
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.7.1/jquery.min.js"></script>
    <script src="javascripts/main.js"></script>
    <!--[if lt IE 9]>
      <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">

  </head>
  <body>

      <header>
        <h1>InPhaDel</h1>
        <p>InPhaDel: (In)tegrating whole genome and proximity ligation sequencing to (Pha)se (Del)etion variants.</p>
      </header>

      <div id="banner">
        <span id="logo"></span>

        <a href="https://github.com/l337x911/inphadel" class="button fork"><strong>View On GitHub</strong></a>
        <div class="downloads">
          <span>Downloads:</span>
          <ul>
            <li><a href="https://github.com/l337x911/inphadel/zipball/master" class="button">ZIP</a></li>
            <li><a href="https://github.com/l337x911/inphadel/tarball/master" class="button">TAR</a></li>
          </ul>
        </div>
      </div><!-- end banner -->

    <div class="wrapper">
      <nav>
        <ul></ul>
      </nav>
      <section>
        <p>Developed by Anand D. Patel in Vineet Bafna's lab at University of California,
San Diego patel [dot] anandd [at] gmail [dot] com</p>

<p>InPhaDel takes as input, a scaffold of phased single nucleotide variants (vcf),
whole genome sequencing data (sorted bam), proximity ligation data (sorted
bam), and deletions (bed) and a classification model ('rf|svm|knn').</p>

<p>InPhaDel returns a prediction for the deletion being homozygous, scaffold A,
scaffold B, or unlikely to be a deletion (bed).  For details of the method, see
Patel, Selvaraj, Bafna 2015.</p>

<h1>
<a id="installing-" class="anchor" href="#installing-" aria-hidden="true"><span class="octicon octicon-link"></span></a>Installing </h1>

<h2>
<a id="software-requirements-" class="anchor" href="#software-requirements-" aria-hidden="true"><span class="octicon octicon-link"></span></a>Software Requirements </h2>

<p>InPhaDel requires Python 2.7.x with packages numpy 1.9, pandas 1.13,
scikit-learn 2.7.6, matplotlib [optional].  This tool may work on other version
of pandas, scikit-learn, and numpy, but has not been tested.</p>

<p>The tool also requires samtools </p>

<h2>
<a id="installing-inphadel-" class="anchor" href="#installing-inphadel-" aria-hidden="true"><span class="octicon octicon-link"></span></a>Installing InPhaDel </h2>

<p>InPhaDel can be run from source or installed.</p>

<p>Download the tar or zip ball and install using the setup.py script.</p>

<pre><code>python setup.py install
</code></pre>

<p>To verify the installation/configuration is correct and run a small test case
for InPhaDel (takes &lt;5min) ::</p>

<pre><code>python setup.py test 
</code></pre>

<p>If all is well, continue to usage.</p>

<p>Alternatively, to run from source directory use ::</p>

<pre><code>python -m svphase.test.predict
</code></pre>

<h1>
<a id="usage-" class="anchor" href="#usage-" aria-hidden="true"><span class="octicon octicon-link"></span></a>Usage </h1>

<p>InPhaDel requires inputs to follow a specific naming convention as described
below,</p>

<p>a) Deletions must be in a simple bed file. Chromosome names must match
reference, and below input bam filenames.</p>

<p>b) Reference in fasta file format (with fai index). Use <code>samtools faidx</code> to
generate index.</p>

<p>c) Phased scaffolds in VCF file format, WGS, and HiC bam files split by
chromosome. Some are required, and optional files will be generated. </p>

<ul>
<li>INPUT_DIR/wgs/all.stat - generated by <code>samtools idxstat</code> of composite wgs
data [required]</li>
<li>INPUT_DIR/hic/all.stat - generated by <code>samtools idxstat</code> of composite hic
data [required]</li>
<li>INPUT_DIR/wgs/CHROMOSOME.all.bam[.bai] - WGS data where .bai is generated
by <code>samtools index</code> [required]</li>
<li>INPUT_DIR/hic/CHROMOSOME.all.bam[.bai] - HiC data where .bai is generated
by <code>samtools index</code> [required]</li>
<li>INPUT_DIR/vcf/CHROMOSOME.vcf [optional if INPUT_DIR/wgs/CHROMOSOME.[AB].bam
and INPUT_DIR/hic/CHROMOSOME.[AB].bam] have been generated.</li>
</ul>

<p>INPUT_DIR is an argument to <code>inphadel</code> and CHROMOSOME is the name of a
chromosome in the reference fasta file. There must be CHROMOSOME bam files for
each CHROMOSOME in the bed file.</p>

<p>To see command line arguments, use ::</p>

<pre><code>inphadel -h
</code></pre>

<p>or from source ::</p>

<pre><code>python -m svphase.inphadel -h
</code></pre>
      </section>
      <footer>
        <p>Project maintained by <a href="https://github.com/l337x911">l337x911</a></p>
        <p><small>Hosted on GitHub Pages &mdash; Theme by <a href="https://twitter.com/michigangraham">mattgraham</a></small></p>
      </footer>
    </div>
    <!--[if !IE]><script>fixScale(document);</script><![endif]-->
    
  </body>
</html>
